common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -366,59 +418,112 @@ private void validateEncodeVersion(String encryptedBlob, int encodeVersionLength
+        if (key == null) {
+            key = createKey();
         }
 
-        return secretKeyData;
+        return key;
     }
 
-    @Override
-    public synchronized SecretKey loadSecretKeyForEncryption(String defaultBlobVersion) throws IOException,
-            GeneralSecurityException {
-        // Loading key only once for performance. If API is upgraded, it will
-        // restart the device anyway. It will load the correct key for new API.
-        if (mKey != null && mHMACKey != null) {
-            return mKey;
+    /**
+     * Load SecretKey for ADAL/MSAL for encryption.
+     * This will try getting key in the following order.
+     * 1. User-defined key.
+     * 2. Keystore-encrypted symmetric key.
+     */
+    private SecretKey loadSecretKeyForADALMSALEncryption() throws IOException, GeneralSecurityException {
+        if (AuthenticationSettings.INSTANCE.getSecretKeyData() != null) {
+            setBlobVersion(VERSION_USER_DEFINED);
+            return loadSecretKey(KeyType.ADAL_USER_DEFINED_KEY);
+        }
+
+        // Try loading existing keystore-encrypted key. If it doesn't exist, create a new one.
+        setBlobVersion(VERSION_ANDROID_KEY_STORE);
+        SecretKey key = loadSecretKey(KeyType.KEYSTORE_ENCRYPTED_KEY);
+
+        if (key == null) {
+            key = createKey();
+        }
+
+        return key;
+    }
+
+    /**
+     * A function for setting mBlobVersion.
+     * Exposed for test cases.
+     */
+    protected void setBlobVersion(@NonNull String blobVersion) {
+        mBlobVersion = blobVersion;
+    }
+
+    /**
+     * Given the key type, load a secret key.
+     */
+    protected SecretKey loadSecretKey(@NonNull final KeyType keyType) throws IOException, GeneralSecurityException {
+        final String methodName = "loadSecretKey";
+
+        switch (keyType) {
+            case LEGACY_AUTHENTICATOR_APP_KEY:
+                Logger.verbose(TAG + methodName, "Loading legacy authApp key.");
+                return getSecretKey(AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME));
+
+            case LEGACY_COMPANY_PORTAL_KEY:
+                Logger.verbose(TAG + methodName, "Loading legacy companyPortal key.");
+                return getSecretKey(AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(COMPANY_PORTAL_APP_PACKAGE_NAME));
+
+            case ADAL_USER_DEFINED_KEY:
+                Logger.verbose(TAG + methodName, "Loading ADAL userDefined key.");
+                return getSecretKey(AuthenticationSettings.INSTANCE.getSecretKeyData());
+
+            case KEYSTORE_ENCRYPTED_KEY:
+                return getKey();
         }
 
-        mBlobVersion = defaultBlobVersion;
-        return getKeyOrCreate(mBlobVersion);
+        Logger.verbose(TAG + methodName, "Unknown KeyType. This code should never be reached.");
+        throw new GeneralSecurityException(ErrorStrings.UNKNOWN_ERROR);
```

Comment:
Could you specify the error code here? Like unable to load the secret key or something else.

Reply:
We could, but given that this code should never be reached, if it happens, our logic is wrong and there is no way to recover from such state. Therefore I don't think introducing another error code would benefit here.